### ISSUE TITLE
RestTemplate response body null if I switch from Boot 1.1.9 to 1.2.0

### ISSUE THREAD

**Issue by khoubyari** [spring-projects/spring-boot#2200]
Hi,

I have simple REST client code that was working with 1.1.x but when I switch to 1.2 the response body becomes null! 

```java
ResponseEntity<String> response =  new RestTemplate().exchange(someURL, HttpMethod.GET, someStringEntity, String.class);
// response comes back with status code 200 and all the correct headers but response.getBody() returns null
```
If I switch back to Spring Boot 1.1.9 then the body is not null and has the correct JSON string.
My POM dependencies look like this (my application is packaged as a war with embedded Tomcat)

```xml

    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>1.2.0.RELEASE</version>
    </parent>

    <dependencies>

        <dependency> <!-- custom logging library; uses GSON -->
            <groupId>com.sample</groupId>
            <artifactId>custom-logging-lib</artifactId>
            <version>0.0.2-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>com.google.code.gson</groupId>
            <artifactId>gson</artifactId>
            <version>2.3.1</version>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-actuator</artifactId>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-security</artifactId>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-tomcat</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-jpa</artifactId>
        </dependency>

        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
            <version>1.4.180</version>
        </dependency>

        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
        </dependency>

        <dependency>  <!-- these are needed for some custom marshalling in the app -->
            <groupId>com.fasterxml.jackson.datatype</groupId>
            <artifactId>jackson-datatype-jdk8</artifactId>
            <version>2.4.4</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.datatype</groupId>
            <artifactId>jackson-datatype-jsr310</artifactId>
            <version>2.4.4</version>
        </dependency>

        <dependency>
            <groupId>com.mangofactory</groupId>
            <artifactId>swagger-springmvc</artifactId>
            <version>0.8.5</version>
        </dependency>
        <dependency>
            <groupId>org.ajar</groupId>
            <artifactId>swagger-spring-mvc-ui</artifactId>
            <version>0.2</version>
        </dependency>

        <dependency>
            <groupId>org.hsqldb</groupId>
            <artifactId>hsqldb</artifactId>
            <scope>runtime</scope>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>com.jayway.jsonpath</groupId>
            <artifactId>json-path</artifactId>
            <version>0.9.1</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>com.jayway.jsonpath</groupId>
            <artifactId>json-path-assert</artifactId>
            <version>0.9.1</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.codehaus.jackson</groupId>
            <artifactId>jackson-core-asl</artifactId>
            <version>1.9.13</version>
        </dependency>

        <dependency>
            <groupId>com.googlecode.json-simple</groupId>
            <artifactId>json-simple</artifactId>
            <version>1.1.1</version>
        </dependency>
        
        <dependency>
		<groupId>org.testng</groupId>
		<artifactId>testng</artifactId>
		<version>6.8.8</version>
	</dependency>
		
	<dependency>
		<groupId>com.jayway.restassured</groupId>
		<artifactId>rest-assured</artifactId>
		<version>2.4.0</version>
	</dependency>

    </dependencies>


```

If I change nothing in my application except change the boot version back to 1.1.9.RELEASE everything works as before. 

Thanks for your help.

**Comment by blimey74** [comment]
I am having the same problem, I have created a sample project to reproduce the problem but I can't reproduce it against a public REST service I've tried only against my company's internal one. My GET call requires http field headers to be set which may or may not be part of the problem.  Is there some publicly available REST services you could recommend that I could try my code against, examples with and without field headers?  Then I'll post my sample project if I can reproduce it against them.

**Comment by khoubyari** [comment]
Sorry I got busy with a release. Still using 1.1.9 until I can get back to this. Anyway if you're looking for a GET service that returns JSON you can try this: 
```
GET http://data.nasa.gov/api/
{"status":"ok","json_api_version":"1.1.1","controllers":["core"]}
```

**Comment by blimey74** [comment]
Hi Siamak, my question was aimed more at the Spring guys.  I've tried the NASA URL and it works fine with Spring 4.1.3.RELEASE.  I'm curious about your code, do you add any Http Headers to your request code? This is how my sample code looks:

```java
       REQUEST_FACTORY.setReadTimeout(10000);
        REQUEST_FACTORY.setConnectTimeout(10000);

        RestTemplate restTemplate = new RestTemplate(REQUEST_FACTORY);

        HttpHeaders requestHeaders = new HttpHeaders();
        requestHeaders.set(FIELDS_REQUEST_HEADER, "default,activationTimes");
        HttpEntity<?> requestEntity = new HttpEntity(requestHeaders);

        ResponseEntity response = restTemplate.exchange("http://data.nasa.gov/api/",
                HttpMethod.GET, requestEntity, String.class, "");


        return (String) response.getBody();
  ```

**Comment by khoubyari** [comment]
@blimey74 : No I did not add a header. Just a simple GET request that expects string as a response.

**Comment by bclozel** [comment]
Hi there,
I think the culprit is [SPR-8016](https://jira.spring.io/browse/SPR-8016) (and me, actually). This change was introduced in Spring Framework 4.1.3 (and is also in 4.1.4) in order to better support empty HTTP body responses with the HTTP RFC.

The problem is this first attempt was incomplete and may have introduced regressions; it looks like it did. A better fix has been pushed and will be available in Spring Framework 4.1.5 and 4.2.

In the meantime, the source of this issue probably has something to do with the HTTP response headers sent by your remote web service (or most likely, the reverse proxy in front of it - varnish, HAProxy?). An easy "fix" would be to make sure that the remote web service sends the `Content-Length` or `Transfer-Encoding: chunked` response header.

**Comment by blimey74** [comment]
@bclozel Hi thanks for the input, I'll try out the workaround and I'll put some feedback up here.

**Comment by khoubyari** [comment]
This problem seems to have been fixed in 1.2.1. All my tests pass. Thank you.

**Comment by blimey74** [comment]
Verified fixed with:

<spring.version>4.1.5.RELEASE</spring.version>
<spring.boot.version>1.2.1.RELEASE</spring.boot.version>

thanks a lot!
